### ISSUE TITLE
Unit tests on Windows are failing depite CI is passing

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#33347]
### Steps to reproduce

Reporting from a discussion among maintainers in Slack. Unit tests running on Windows are failing, despite CI on them passing. Examples of failing unit tests with :heavy_check_mark: CI:
- https://github.com/spack/spack/actions/runs/3246813188/jobs/5326063183
- https://github.com/spack/spack/actions/runs/3246813188/jobs/5326063112

The issue making unit tests fail possibly started in https://github.com/spack/spack/pull/30690, though we also need to fix CI scripts so that next time they report on the error immediately.

### Error message

![Screenshot from 2022-10-14 18-39-56](https://user-images.githubusercontent.com/4199709/195978555-d7f631f8-979e-45b0-b791-716062810105.png)

### Information on your system

NA (It happens on Github Actions)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
@johnwparent I know you're working on this (thanks!). Added the issue so that we can keep track of it until it's solved and add details that might be relevant without them being subject to Slack's limitations.

**Comment by johnwparent** [comment]
Thanks @alalazo! I've also added this to the Spack on Windows project.

**Comment by alalazo** [comment]
I'm not a Windows user, but in case it helps, it seems the default GithubActions behavior for powershell is:
- Prepend `$ErrorActionPreference = 'stop'` to the script being executed
- Append `if ((Test-Path -LiteralPath variable:\LASTEXITCODE)) { exit $LASTEXITCODE }` to reflect the last exit status

Information taken from [here](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#custom-shell).

EDIT: there's also this [SO question](https://stackoverflow.com/questions/9948517/how-to-stop-a-powershell-script-on-the-first-error) with some more details.

**Comment by johnwparent** [comment]
The last exit status is actually the issue. The `$ErrorActionPreference` only applies to poweshell commandlets unfortunately, not to outside exes like `python ...` or `coverage`. So while the spack unit-test command fails, the next two commands (the coverage commands) still exit zero, so when the `if ((Test-Path -LiteralPath variable:\LASTEXITCODE)) { exit $LASTEXITCODE}` is run, it sees the exit code from coverage and passes.

There are two ways to solve this. We can wrap every spack command executed on the CI in a cmdlet or we can check the $LASTEXITCODE after each command. I elected for the second approach in the PR I just linked to this issue.
